docs/blog/2021-06-16-tailor-your-slack-audit-sink-messages.md

```diff
@@ -0,0 +1,149 @@
+
+```go
+reqJSON, err := protojson.Marshal(event.RequestMetadata.Body)
+respJSON, err := protojson.Marshal(event.ResponseMetadata.Body)
+```
+
+The aim here is to unpack the metadata (Any type) into its underlying message. We looked into using anypb’s [UnmarshalNew](https://pkg.go.dev/google.golang.org/protobuf/types/known/anypb#Any.UnmarshalNew) but ultimately decided upon protojson’s [Marshal](https://pkg.go.dev/google.golang.org/protobuf/encoding/protojson#Marshal). The main reason for this choice was that we wanted to use the json format as the basis for the context in order to simplify the metadata structure a bit. For example, protojson flattens oneOf fields into its top level field. Since custom messages need to specify what exact data is needed from the request/response, simplifications like these can help make writing a custom message simpler/easier.
+
+```go
+var requestMetadata map[string]interface{}
+var responseMetadata map[string]interface{}
+
+json.Unmarshal(reqJSON, &requestMetadata)
+json.Unmarshal(respJSON, &responseMetadata)
+```
+
+API request/response messages can be complex types, so we unmarshaled the metadata into a `map[string]interface{}`. This form can be used with many other packages and in particular we wanted to support any of the Golang template [functions](https://golang.org/pkg/text/template/#hdr-Functions) (and make it easy for contributors to enable template helpers like [Sprig](http://masterminds.github.io/sprig/)) so that users can easily retrieve the metadata for their custom Slack audits.
+
+```go
+return &auditTemplateData{
+	Request:  requestMetadata,
+	Response: responseMetadata,
+
+}
+```
+
+Lastly, we didn’t want the users to need to know anything about the schema of the audit event metadata, so we saved the metadata in a new simple struct. In this way, the user can chain the information they wanted to retrieve with just `.Request.<field name>` or `.Response.<field name>`.
+
+## Conclusion
+
+The Clutch team values building features generically as possible so that it can serve many current and future use cases. We didn’t want workflow admins/creators to need to add custom logic to the audit architecture per Clutch workflow to support this feature. Our aim was for it to be intuitive for someone to use without needing to know what’s happening under the hood (like the specifics of the audit flow or Slack sink).
+
+What’s next? In next iterations, we’d like to move the custom messages alongside API code so that there’s less configuration in the clutch-config and experiment with creating a default customized Slack message for workflows. Also, a new feature we’d like to add is targeted Slack notifications, so that you can route your audits to various channels in your Slack workspace.
```

Comment:
```suggestion
What’s next? In future iterations, we’d like to move the custom messages alongside API code so that there’s less configuration in the clutch-config and experiment with creating a standard customized Slack message for workflows. We would also like to add targeted Slack notifications, so that you can route your audits to various channels in your Slack workspace.
```

Reply:
Should we create tickets for the future enhancements & features and link them here in case someone wants to work on them?

Reply:
good idea, linked the issues